neurodamus/core/_utils.py

```diff
@@ -67,8 +68,39 @@ def rank0_wrapper(*args, **kw):
         return rank0_wrapper
 
 
-def return_neuron_timings(f):
-    """Decorator to collect and return timings on a neuron run
+class SimulationProgress:
+    def __init__(self):
+        """
+        Utility class which will set up a timer to perioducally check the amount of time lapsed
+        in the simulation compared to the final tstop value.  This is converted into a percentage
+        of the job complete which is then printed to the console.
+        """
+        self.last_time_check = time.time()
+        self.sim_start = self.last_time_check
+        self.update_progress()
+
+    @run_only_rank0
+    def update_progress(self):
+        """
+        Callback function that refreshes the progress value (if enough time has elapsed) and then
+        inserts the next call into the event queue.
+        """
+        current_time = time.time()
+        sim_t = Nd.t
+        sim_tstop = Nd.tstop
+        if (current_time - self.last_time_check > 0.75) and (sim_t > 0):
+            self.last_time_check = current_time
+            sec_remain = (self.last_time_check - self.sim_start) * (sim_tstop / sim_t - 1)
+            hours = sec_remain / 3600
+            minutes = (sec_remain % 3600) / 60
+            seconds = (sec_remain % 3600) % 60
+            print(f"\r[t={sim_t:5.2f}] Completed {sim_t*100/sim_tstop:2.0f}% "
+                  f"ETA: {hours:02.0f}:{minutes:02.0f}:{seconds:02.0f}  ", end='', flush=True)
+        Nd.cvode.event(sim_t + 1, self.update_progress)
+
+
+def return_neuron_timings_and_progress(f):
```

Comment:
I wouldn't rename this - you don't return the progress?

Progress could be a separate decorator, too.